src/test/CharterManager-unit.t.sol

```diff
@@ -0,0 +1,601 @@
+        b.allow(address(a));
+        a.frob(address(b), address(b), address(a), 100 * 1e18, 50 * 1e18);
+        assertEq(b.gems(), 0);
+        (uint256 ink, uint256 art) = b.urn();
+        assertEq(ink, 100 * 1e18);
+        assertEq(art, 50 * 1e18);
+        assertEq(a.dai(), 50 * 1e45);
+        assertEq(a.gems(), 0);
+        a.frob(address(b), address(b), address(a), -100 * 1e18, -50 * 1e18);
+        (ink, art) = b.urn();
+        assertEq(ink, 0);
+        assertEq(art, 0);
+        assertEq(a.dai(), 0);
+        assertEq(b.gems(), 100 * 1e18);
+    }
+
+    function testFail_frob_other1() public {
+        (Usr a, Usr b) = init_user();
+        a.join(address(b), 100 * 1e6);
+        a.frob(address(b), address(b), address(a), 100 * 1e18, 50 * 1e18);
+    }
+
+    function testFail_frob_other2() public {
+        (Usr a, Usr b) = init_user();
+        a.join(address(b), 100 * 1e6);
+        b.allow(address(a));
+        b.disallow(address(a));
+        a.frob(address(b), address(b), address(a), 100 * 1e18, 50 * 1e18);
+    }
+
+    function test_flux_to_other() public {
+        (Usr a, Usr b) = init_user();
+        a.join(100 * 1e6);
+        assertEq(a.gems(), 100 * 1e18);
+        a.flux(address(a), address(b), 100 * 1e18);
+        assertEq(b.gems(), 100 * 1e18);
+        b.exit(100 * 1e6);
+        assertEq(b.gems(), 0);
+        assertEq(gem.balanceOf(address(b)), 300 * 1e6);
+    }
+    function test_flux_yourself() public {
+        // Flux to yourself should be a no-op
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        assertEq(a.gems(), 100 * 1e18);
+        a.flux(address(a), address(a), 100 * 1e18);
+        assertEq(a.gems(), 100 * 1e18);
+        a.exit(100 * 1e6);
+        assertEq(a.gems(), 0);
+        assertEq(gem.balanceOf(address(a)), 200 * 1e6);
+    }
+
+    function test_flux_from_other() public {
+        (Usr a, Usr b) = init_user();
+        a.join(100 * 1e6);
+        assertEq(a.gems(), 100 * 1e18);
+        a.allow(address(b));
+        b.flux(address(a), address(b), 100 * 1e18);
+        assertEq(a.gems(), 0);
+        assertEq(b.gems(), 100 * 1e18);
+        b.exit(100 * 1e6);
+        assertEq(b.gems(), 0);
+        assertEq(gem.balanceOf(address(b)), 300 * 1e6);
+    }
+
+    function testFail_flux_from_other1() public {
+        (Usr a, Usr b) = init_user();
+        a.join(100 * 1e6);
+        b.flux(address(a), address(b), 100 * 1e18);
+    }
+
+    function testFail_flux_from_other2() public {
+        (Usr a, Usr b) = init_user();
+        a.join(100 * 1e6);
+        a.allow(address(b));
+        a.disallow(address(b));
+        b.flux(address(a), address(b), 100 * 1e18);
+    }
+
+    function testFail_quit() public {
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        a.frob(100 * 1e18, 50 * 1e18);
+        a.quit();       // Attempt to unbox the urn (should fail when vat is live)
+    }
+    function test_quit() public {
```

Comment:
I think for all of these tests that call vat functions we will want to `hevm.warp` to the future, drip the collateral, then attempt to make the calls to ensure we're properly calculating `dink` and `dart`.

Reply:
Moved to use dss contracts instead of mock contracts and added a test fo drip and withdraw (test_drip_withdraw).
Adding rates accumulation to all of the tests will very much complicate dai and fees calculations so prefer to keep it to this test.